### ISSUE TITLE
@Scheduled task instrumentation does not work for Kotlin suspend functions

### ISSUE THREAD

**Issue by pwestlin** [spring-projects/spring-framework#32165]
[Spring docs for Scheduled tasks instrumentation](https://docs.spring.io/spring-framework/reference/integration/observability.html#observability.tasks-scheduled) states:

> An Observation is created for [each execution of an @Scheduled task](https://docs.spring.io/spring-framework/reference/integration/scheduling.html#scheduling-enable-annotation-support).

This function get an automatic observation:
```kotlin
@Scheduled(fixedDelay = 5, timeUnit = TimeUnit.SECONDS)
fun nonSuspendable() {
    logger.info("Not suspendable")
}
```
but this **suspend** function does not:
```kotlin
@Scheduled(fixedDelay = 5, timeUnit = TimeUnit.SECONDS)
suspend fun suspendable() {
    logger.info("Suspendable")
}
```

I use **Spring Boot 3.2.2** and I've also tried 3.2.3-SNAPSHOT and 3.3.0-M1.

build.gradle.kts (shortened):
```yaml
plugins {
    id("org.springframework.boot") version "3.2.2"
    id("io.spring.dependency-management") version "1.1.4"
    kotlin("jvm") version "1.9.22"
    kotlin("plugin.spring") version "1.9.22"
}

java {
    sourceCompatibility = JavaVersion.VERSION_21
}

repositories {
    mavenCentral()
}

dependencies {
    implementation("org.springframework.boot:spring-boot-starter-webflux")
    implementation("org.springframework.boot:spring-boot-starter-actuator")
    implementation("com.fasterxml.jackson.module:jackson-module-kotlin")
    implementation("io.projectreactor.kotlin:reactor-kotlin-extensions")
    implementation("org.jetbrains.kotlin:kotlin-reflect")
    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-reactor")

    implementation("io.micrometer:micrometer-tracing")
    implementation("io.micrometer:micrometer-tracing-bridge-brave")

    testImplementation("org.springframework.boot:spring-boot-starter-test")
    testImplementation("io.projectreactor:reactor-test")
}
```

Application code:
```kotlin
@SpringBootApplication
@EnableScheduling
class Application

fun main(args: Array<String>) {
    Hooks.enableAutomaticContextPropagation()
    runApplication<Application>(*args)
}

@Service
class SchedulingService {

    private val logger: Logger = LoggerFactory.getLogger(this.javaClass)

    @Scheduled(fixedDelay = 5, timeUnit = TimeUnit.SECONDS)
    fun nonSuspendable() {
        logger.info("Not suspendable")
    }

    @Scheduled(fixedDelay = 5, timeUnit = TimeUnit.SECONDS)
    suspend fun suspendable() {
        logger.info("Suspendable")
    }
}
```

When I run the application I get the following output:
```
2024-01-31T07:33:26.925+01:00  INFO [65b9e9b65a4ecd1702feecf2dbdd6be4,02feecf2dbdd6be4] 926648 --- [   scheduling-1] [65b9e9b65a4ecd1702feecf2dbdd6be4-02feecf2dbdd6be4] n.w.s.SchedulingService                  : Not suspendable
2024-01-31T07:33:26.946+01:00  INFO [,] 926648 --- [   scheduling-1] [                                                 ] n.w.s.SchedulingService                  : Suspendable
```
There we can se that "65b9e9b65a4ecd1702feecf2dbdd6be4,02feecf2dbdd6be4" means that the function "nonSuspendable" gets an observation but "suspendable" doesn't.

Regards Peter

**Comment by bclozel** [comment]
@sdeleuze and I investigated this issue and we've found that this is not a simple Spring Framework issue, but rather a broader problem with Kotlin Coroutines and Observability. This issue explored in micrometer-metrics/tracing#174.

## How this works with plain Spring MVC

With Spring MVC, when an observation is created, we can then use it to open a scope around some code execution. The opening of this scope is handled by the tracing infrastructure and sets up the relevant ThreadLocal and MDC values. The logging statements, when executed, have all the information in context and you can see traceId and spanId.

## How this works with Reactor Mono and Flux

With Reactor, work can be scheduled on any worker thread so this is not as straightforward.
Reactor is using the [Micrometer context-propagation](https://github.com/micrometer-metrics/context-propagation) library to propagate the information across contexts.
Here, our instrumentation is putting the current observation in the Reactor `ContextView` under the `ObservationThreadLocalAccessor.KEY` key. If the context propagation feature is enabled (typically in Spring Boot with `spring.reactor.context-propagation=auto`, then Reactor uses the registered context propagation ThreadLocalAccessor to get the information from the `ContextView` and temporarily restore it as ThreadLocal in the current scope. The `ObservationThreadLocalAccessor` takes care of opening and closing the observation scope, which sets up the MDC as expected.

## With Kotlin Coroutines

In this case, the reactor context is propagated by the `org.jetbrains.kotlinx:kotlinx-coroutines-reactor` library automatically, but under a specific key. Unlike Reactor's `ContextView`, keys are not String-based but rather must implement the `CoroutineContext.Key` interface. This means that you can get the entire Reactor context from the current coroutine context and get values from it.

But unlike Reactor, there is no automatic integration with context propagation, ThreadLocals or the MDC. In fact, it seems Kotlin Coroutines expect users to directly interact with the context to get values, or compose with a `CoroutineContext` implementation that implements `ThreadContextElement` to set and restore context values.

I have managed to get this working with a custom function that leverages existing Micrometer infrastructure:

```java
import io.micrometer.core.instrument.kotlin.asContextElement
import io.micrometer.observation.Observation
import io.micrometer.observation.contextpropagation.ObservationThreadLocalAccessor
import kotlinx.coroutines.reactor.ReactorContext
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import reactor.util.context.ContextView
import java.util.concurrent.TimeUnit
import kotlin.coroutines.CoroutineContext
import kotlin.coroutines.coroutineContext

@Component
class SchedulingService {

    companion object {
        val logger = LoggerFactory.getLogger(SchedulingService::class.java.name)
    }

    @Scheduled(fixedDelay = 5, timeUnit = TimeUnit.SECONDS)
    suspend fun suspendable() {
        withContext(observationContext(coroutineContext)) {
            logger.info("Suspendable")
        }
    }

    fun observationContext(context: CoroutineContext) : CoroutineContext {
        // get the Reactor context from the CoroutineContext
        val contextView = context[ReactorContext]!!.context as ContextView
        // this context contains the current observation under this well known key
        // because the @Scheduled instrumentation contributed it
        val observation = contextView.get(ObservationThreadLocalAccessor.KEY) as Observation
        // we can then use this Micrometer context to wrap the execution
        // the observation scope and MDC will be taken care of
        return observation.observationRegistry.asContextElement()
    }
}
```

This is by no means the solution we're advertizing and we're not sure how to tackle this problem at this point.
I'm personally not convinced that we should restore ThreadLocals and MDC transparently for developers, as the kotlinx library does not do this in the first place.

I'm leaving this issue opened for now because we might want to revisit the Coroutine to Publisher arrangement in this case, in order to pass a custom CoroutineContext. But this issue should be mainly discussed and tackled with the Micrometer team.

**Comment by bclozel** [comment]
I think I've refined a solution that could be contributed to the context-propagation project.

This `ThreadContextElement` will help Kotlin apps to lift the `ReactorContext` from the `CoroutineContext` if it's there, or just capture the context from the local `ThreadLocals`:

```kotlin
package io.mircrometer.context

import io.micrometer.context.ContextRegistry
import io.micrometer.context.ContextSnapshot
import io.micrometer.context.ContextSnapshotFactory
import kotlinx.coroutines.ThreadContextElement
import kotlinx.coroutines.reactor.ReactorContext
import reactor.util.context.ContextView
import kotlin.coroutines.AbstractCoroutineContextElement
import kotlin.coroutines.CoroutineContext


class PropagationContextElement(private val context: CoroutineContext) : ThreadContextElement<ContextSnapshot.Scope>,
    AbstractCoroutineContextElement(Key) {


    public companion object Key : CoroutineContext.Key<PropagationContextElement>

    val contextSnapshot: ContextSnapshot
        get() {
            val contextView: ContextView? = context[ReactorContext]?.context
            val contextSnapshotFactory =
                ContextSnapshotFactory.builder().contextRegistry(ContextRegistry.getInstance()).build()
            if (contextView != null) {
                return contextSnapshotFactory.captureFrom(contextView)
            }
            return contextSnapshotFactory.captureAll()
        }

    override fun restoreThreadContext(context: CoroutineContext, scope: ContextSnapshot.Scope) {
        scope.close()
    }

    override fun updateThreadContext(context: CoroutineContext): ContextSnapshot.Scope {
        return contextSnapshot.setThreadLocals()
    }
}
```

Using it in the application leverages the context-propagation `ThreadLocalAccessors` as Reactor does and does not need to depend on Micrometer Observation or MDC.

```
    @Scheduled(fixedDelay = 5, timeUnit = TimeUnit.SECONDS)
    suspend fun suspendable() {
        withContext(PropagationContextElement(coroutineContext)) {
            logger.info("Suspendable")
        }
    }
```

WDYT @sdeleuze ?

**Comment by sdeleuze** [comment]
I think I like to direction taken by this proposal as Spring support for Coroutines is tightly linked to the Reactive support, and the scope of this issue is probably wider than scheduling support and could potentially solved what has been discussed for months in https://github.com/micrometer-metrics/tracing/issues/174.

But we should experiment and discuss about where and how this potential feature should be provided and integrated. To make this feature usable, I think we should find a way to configure it automatically.

As proposed by @antechrestos in [this comment](https://github.com/spring-projects/spring-framework/issues/31893#issuecomment-1907576756), I am wondering if we could use this kind of facility to provide that support seamlessly at Spring level.

For example in `org.springframework.web.reactive.result.method.InvocableHandlerMethod`, should we change:
```kotlin
if (coroutineContext == null) {
    return CoroutinesUtils.invokeSuspendingFunction(method, target, args);
}
else {
    return CoroutinesUtils.invokeSuspendingFunction((CoroutineContext) coroutineContext, method, target, args);
}
````

To something like:
```kotlin
if (coroutineContext == null) {
    return CoroutinesUtils.invokeSuspendingFunction(PropagationContextElement(Dispatchers.getUnconfined()), method, target, args);
}
else {
    return CoroutinesUtils.invokeSuspendingFunction(PropagationContextElement((CoroutineContext) coroutineContext), method, target, args);
}
```
and do the same for `ScheduledAnnotationReactiveSupport`?

IMO that could provide the level of integration people expect from Spring and would be consistent with what we do on Reactive side.

If provided at context-propagation level, we should make sure this will be provided in a dependency we can use in Spring Java code (potentially using defensive classpath check and nest class to ensure this remains optional).

Any thoughts?

**Comment by ilya40umov** [comment]
@sdeleuze btw, I guess you meant to write something like this (you need to use `plus` operation to "mix in" the new element):
```
var context = coroutineContext != null ? (CoroutineContext) coroutineContext : Dispatchers.getUnconfined();
return CoroutinesUtils.invokeSuspendingFunction(context.plus(PropagationContextElement(context)), method, target, args);
```
---

Either way, I would totally love having something like `PropagationContextElement` at our disposal, as right now the only option to propagate the observation (and thus the trace) is via [KotlinObservationContextElement](https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/instrument/kotlin/KotlinObservationContextElement.java) from Micrometer Observation library. 

However, `KotlinObservationContextElement` only interacts with `ObservationThreadLocalAccessor`, and thus does nothing to handle [ObservationAwareSpanThreadLocalAccessor](https://github.com/micrometer-metrics/tracing/blob/main/micrometer-tracing/src/main/java/io/micrometer/tracing/contextpropagation/ObservationAwareSpanThreadLocalAccessor.java) from Micrometer Tracing. 

If Spring would then add this element to the context in all cases when suspending Kotlin code is called (i.e. WebFlux Controller methods, coRouter, @Scheduled etc.) it would hopefully take care of all ThreadLocalAccessors at once.

I am wondering if `KotlinObservationContextElement` would also work for the cases when a child observation / span etc. is created and needs to be put into the scope. E.g. I'm currently trying to add some more Kotlin-friendly API to Observation in in https://github.com/micrometer-metrics/micrometer/pull/4823, and one of the main use-cases for this API is to be able to create a nested observation (e.g. to have a new span to demarcate a section of the code).

**Comment by bclozel** [comment]
I have discussed this with the Micrometer team and finding a proper home for `PropagationContextElement` is not that easy. This class depends on `io.micrometer:context-propagation` and `org.jetbrains.kotlinx:kotlinx-coroutines-reactor` (which itself depends on kotlin and `io.projectreactor:reactor-core`).

* we can't add that to the context-propagation project, as this would create a dependency cycle with reactor
* micrometer-core is not a good option either, because you can use this without Micrometer: its main goal is to support context-propagation in Kotlin coroutines and reactor
* maybe `org.jetbrains.kotlinx:kotlinx-coroutines-reactor` would be the best option here, but I'm not sure this project is interested in providing context-propagation support.
* Spring Framework is also an option, but this would significantly restrict the usage of this type: only Spring apps would be supported

**Comment by grassehh** [comment]
Hello, I seem to have found another context propagation issue when using proxied beans.
For example suppose you have the following controller:
```kotlin
 @GetMapping("test")
    @ResponseStatus(OK
    suspend fun test() {
        CoroutineScope(SupervisorJob()).launch(observationRegistry.asContextElement()) {
            proxiedService.doSomethingSuspend()
        }
    }
``` 

And `proxiedService` bean is a simple service proxied using an `AbstractAutoProxyCreator` that just makes a `WebClient` call:
```kotlin
class ProxiedServiceImpl(private val webClient: WebClient) : ProxiedService {
    override suspend fun doSomethingSuspend() {
        webClient.get()
            .uri("https://www.google.fr")
            .awaitExchange {
                println(MDC.getCopyOfContextMap())
            }
    }
}
```
Then the MDC context will be empty in the `awaitExchange`.
To make it work, I need to pass the whole parent coroutine context to the child coroutine instead of just `observationRegistry.asContextElement()`.

If the service is not a proxy, it workds fine. Also if the service is not called inside the coroutine, but rather directly inside the controller method, it works with the proxy too.

So I wonder if there is not an issue with Spring AOP with observability.

**Comment by ilya40umov** [comment]
@grassehh I'm just curious if this is because `CoroutineScope(SupervisorJob()).launch()` will basically launch a new coroutine detached from the one created for handling of HTTP request, which potentially would mean that handling of HTTP request can finish before the code in the launch block ever has chance to execute (and thus the associated observation is potentially stopped by that point).

Could you try grabbing the `job` returned by the `launch` call and calling `join()` on it?

**Comment by grassehh** [comment]
> @grassehh I'm just curious if this is because `CoroutineScope(SupervisorJob()).launch()` will basically launch a new coroutine detached from the one created for handling of HTTP request, which potentially would mean that handling of HTTP request can finish before the code in the launch block ever has chance to execute (and thus the associated observation is potentially stopped by that point).
> 
> Could you try grabbing the `job` returned by the `launch` call and calling `join()` on it?

Doesn't seem to work. You can checkout my sample [here](https://github.com/grassehh/spring-boot-3-tracing-coroutine/blob/a4a40f090268d9d2201b6b4527cc1739aaa30be5/app/src/main/kotlin/com/grassehh/app/controller/SampleController.kt#L73).

The thing is that if you use the `nonProxiedService` instead of `proxiedService` and call the `/aop` route, it works (the MDC context will be preserved).

**Comment by bclozel** [comment]
I'm closing this issue for now, in favor of https://github.com/Kotlin/kotlinx.coroutines/issues/4187.
At this point I think we should explore a `org.jetbrains.kotlinx:kotlinx-coroutines-reactor` integration, as Spring Framework would only serve a fraction of the Kotlin community for this need.

We can reopen this issue if we need to reconsider.

**Comment by bclozel** [comment]
This has been declined in https://github.com/Kotlin/kotlinx.coroutines/issues/4187, so our current guidance is to [copy/paste the relevant code in your project](https://github.com/spring-projects/spring-framework/issues/32165#issuecomment-1920986138). We could consider hosting this type in Spring Framework, assuming we can find a proper place for this; it would need to work with our dependency arrangement and within a module that fits with community usage. We'll see if there is enough demand for this.

**Comment by sdeleuze** [comment]
FYI https://github.com/spring-projects/spring-framework/issues/35485, to be available in Spring Framework 7.0.0-RC1, should make `@Scheduled` + coroutines observability working out of the box thanks to `PropagationContextElement` that @bclozel introduced.
